src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/CSharp/RequiredAttributeParser.cs

```diff
@@ -1,305 +1,373 @@
+            SkipWhitespace();
 
-            ParseCssAttributeName(attributeBuilder);
+            name = ParseCssAttributeName();
 
-            PassOptionalWhitespace();
+            value = null;
+            valueComparison = default;
 
-            if (!EnsureNotAtEnd(attributeBuilder))
+            SkipWhitespace();
+
+            if (!EnsureNotAtEnd())
             {
                 return false;
             }
 
-            if (!TryParseCssValueComparison(attributeBuilder, out RequiredAttributeValueComparison valueComparison))
+            if (!TryParseCssValueComparison(out valueComparison))
             {
                 return false;
             }
 
-            PassOptionalWhitespace();
+            SkipWhitespace();
 
-            if (!EnsureNotAtEnd(attributeBuilder))
+            if (!EnsureNotAtEnd())
             {
                 return false;
             }
 
-            if (valueComparison != RequiredAttributeValueComparison.None && !TryParseCssValue(attributeBuilder))
+            if (valueComparison != RequiredAttributeValueComparison.None &&
+                !TryParseCssValue(out value))
             {
                 return false;
             }
 
-            PassOptionalWhitespace();
+            SkipWhitespace();
 
             if (At(']'))
             {
                 // Move past the ending bracket.
-                _index++;
+                MoveNext();
                 return true;
             }
+
             else if (AtEnd)
             {
-                var diagnostic = RazorDiagnosticFactory.CreateTagHelper_CouldNotFindMatchingEndBrace(_requiredAttributes);
-                attributeBuilder.Diagnostics.Add(diagnostic);
+                _diagnostics.Append(
+                    RazorDiagnosticFactory.CreateTagHelper_CouldNotFindMatchingEndBrace(_input));
             }
             else
             {
-                var diagnostic = RazorDiagnosticFactory.CreateTagHelper_InvalidRequiredAttributeCharacter(Current, _requiredAttributes);
-                attributeBuilder.Diagnostics.Add(diagnostic);
+                _diagnostics.Append(
+                    RazorDiagnosticFactory.CreateTagHelper_InvalidRequiredAttributeCharacter(Current, _input));
             }
 
             return false;
         }
 
-        private bool EnsureNotAtEnd(RequiredAttributeDescriptorBuilder builder)
+        private string ParseCssAttributeName()
         {
-            if (AtEnd)
-            {
-                var diagnostic = RazorDiagnosticFactory.CreateTagHelper_CouldNotFindMatchingEndBrace(_requiredAttributes);
-                builder.Diagnostics.Add(diagnostic);
+            var nameEndIndex = _span.IndexOfAny(s_invalidCssAttributeNameCharacters);
 
-                return false;
+            if (nameEndIndex == -1)
+            {
+                nameEndIndex = _span.Length;
             }
 
-            return true;
-        }
+            var result = _span[..nameEndIndex].ToString();
```

Comment:
>_span[..nameEndIndex].ToString(); [](http://example.com/codeflow?start=25&length=33)

ToString seems unfortunate, but it looks like it tracks back to RequiredAttributeDescriptorBuilder.Name being a string

Reply:
Ultimately, we need to make strings that make up tag helper objects. In this PR, I tried to only produce strings that would be assigned into the builder. Did you have something in mind for making the builders use spans somehow?

Reply:
Nothing specific in mind, probably more of a general lament that most things at the core level are strings.